v2/component/launcher.go

```diff
@@ -441,3 +511,181 @@ func (l *Launcher) RunComponent(ctx context.Context, cmd string, args ...string)
 
 	return l.metadataClient.PublishExecution(ctx, execution, outputParameters, outputArtifacts)
 }
+
+func getRuntimeArtifactSchema(rta *pipeline_spec.RuntimeArtifact) (string, error) {
+	switch t := rta.Type.Kind.(type) {
+	case *pipeline_spec.ArtifactTypeSchema_InstanceSchema:
+		return t.InstanceSchema, nil
+	case *pipeline_spec.ArtifactTypeSchema_SchemaTitle:
+		return "title: " + t.SchemaTitle, nil
+	case *pipeline_spec.ArtifactTypeSchema_SchemaUri:
+		return "", fmt.Errorf("SchemaUri is unsupported, found in RuntimeArtifact %+v", rta)
+	default:
+		return "", fmt.Errorf("unknown type %T in RuntimeArtifact %+v", t, rta)
+	}
+}
+
+func mergeRuntimeArtifacts(src, dst *pipeline_spec.RuntimeArtifact) {
+	if len(src.Uri) > 0 {
+		dst.Uri = src.Uri
+	}
+
+	if src.Metadata != nil {
+		if dst.Metadata == nil {
+			dst.Metadata = src.Metadata
+		} else {
+			for k, v := range src.Metadata.Fields {
+				dst.Metadata.Fields[k] = v
+			}
+		}
+	}
+}
+
+// TODO(neuromage): Move these helper functions to a storage package and add tests.
+func uploadFile(ctx context.Context, bucket *blob.Bucket, localFilePath, blobFilePath string) error {
+	errorF := func(err error) error {
+		return fmt.Errorf("uploadFile(): unable to complete copying %q to remote storage %q: %w", localFilePath, blobFilePath, err)
```

Comment:
Thanks the tip. However, I prefer to avoid named returns except in small functions: go/go-style/decisions#named-result-parameters

I'd prefer the explicit route for readability here as well. Even the comment you linked to starts  with the reader being confused by how  it works :-)

Reply:
That's a fair point!
I did find this hard to understand from the beginning too.

Just to be clear, even when using named return values, we can still write return values explicitly and I also prefer writing them explicitly like you.

```
return nil, err
```

Maybe people will get familiar with this, or maybe not. I am OK with not doing this right now,

but I still think it has some value, because now reading a single defer block, I will know for sure that every error is wrapped consistently with the same info. If we write an error helper, I may forget to call it in some return err cases.